### PR TITLE
8318826: C2: "Bad graph detected in build_loop_late" with incremental inlining

### PR DESCRIPTION
**Pull request by rwestrel**
In test case, `test1()`, `inlined1()` returns null which is used as
receiver for the `notInlined()` virtual call. On the fallthrough
projection of the call, the receiver is casted to not null. In the
case of `test1()` the fallthrough should be unreachable:
`CatchNode::Value()` makes that path dead if it is a projection of a
virtual call with a null receiver. When `test1()` is compiled with
incremental inlining, before `inlined1()` is inlined, the receiver of
the virtual call is not known to be null so the fallthrough path is
not killed. When it is inlined, the first input argument to the
virtual call node is updated to be null. So that node is pushed on the
igvn work list. But the `CatchNode` that's not change in the the
process is not. As a result, the fallthrough path of the virtual call
is not killed. The input to the cast node in the fallthrough path is
changed to null which causes that node to become top. That causes the
crash.

IGVN has logic to enqueue nodes that are related to an edge change but
are not directly modified. We would need similar logic to be called
from late inlining code. This change refactors the igvn logic so it
can be called from elsewhere.

`test2()` is similar to `test1()` except the return of the inlined
method `inlined2()` is not set to null. Instead, a replaced node is
recorded from `a` to null (because class `D` is unrelated to `A` so
casting a `A` to `D` successfully only happens for null) and applied
when `inlined2()` is late inlined.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8318826](https://bugs.openjdk.org/browse/JDK-8318826): C2: "Bad graph detected in build_loop_late" with incremental inlining (**Bug** - P3)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16581/head:pull/16581` \
`$ git checkout pull/16581`

Update a local copy of the PR: \
`$ git checkout pull/16581` \
`$ git pull https://git.openjdk.org/jdk.git pull/16581/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16581`

View PR using the GUI difftool: \
`$ git pr show -t 16581`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16581.diff">https://git.openjdk.org/jdk/pull/16581.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16581#issuecomment-1803857535)

### CODE REVIEW

File: test/hotspot/jtreg/compiler/inlining/TestNullAtCallAfterLateInline.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * bug 8318826
+ * @summary C2: "Bad graph detected in build_loop_late" with incremental inlining
+ * @requires vm.compiler2.enabled
+ * @run main/othervm -XX:+AlwaysIncrementalInline -XX:-BackgroundCompilation TestNullAtCallAfterLateInline
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/inlining/TestNullAtCallAfterLateInline.java:29]: `AlwaysIncrementalInline` is a debug option.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me! Always surprising that these issues went unnoticed for such a long time.

File: test/hotspot/jtreg/compiler/inlining/TestNullAtCallAfterLateInline.java (hunk 1 above)
Comment (vnkozlov) [reply test/hotspot/jtreg/compiler/inlining/TestNullAtCallAfterLateInline.java:29]: May be we should make it diagnostic and add it to stress testing flags.
Or, as Tobias pointed, `@requires` for debug VM is needed here.

Comment (rwestrel) [reply test/hotspot/jtreg/compiler/inlining/TestNullAtCallAfterLateInline.java]: Thanks for looking at this. I added `-XX:+IgnoreUnrecognizedVMOptions` so the test case is run in any case.

File: src/hotspot/share/opto/phaseX.hpp [hunk 2]
```diff
@@ -528,8 +528,9 @@ class PhaseIterGVN : public PhaseGVN {
   }
 
   // Add users of 'n' to worklist
-  void add_users_to_worklist0( Node *n );
-  void add_users_to_worklist ( Node *n );
+  static void add_users_to_worklist0(Node* n, Unique_Node_List& worklist);
+  static void add_users_of_use_to_worklist(Node* n, Node* use, Unique_Node_List& worklist);
+  void add_users_to_worklist (Node *n);
```
Comment (chhagedorn) [inline src/hotspot/share/opto/phaseX.hpp:533]: While at it:
```suggestion
  void add_users_to_worklist(Node* n);
```

File: src/hotspot/share/opto/replacednodes.cpp [hunk 3]
```diff
@@ -222,6 +222,7 @@ void ReplacedNodes::apply(Compile* C, Node* ctl) {
     }
 
     // Clone nodes and record mapping from current to cloned nodes
+    uint old_nodes = C->unique();
```
Comment (chhagedorn) [inline src/hotspot/share/opto/replacednodes.cpp:225]: I suggest the following name instead to better describe what the uint is:
```suggestion
    uint idx_before_clone = C->unique();
```

**Review by chhagedorn — APPROVED** [review]
Looks good to me, too!

**Review by vnkozlov — APPROVED** [review]
Good.

**Review by chhagedorn — APPROVED** [review]
Still good, thanks for the update.

**Comment by rwestrel** [comment]
Thanks for the reviews @vnkozlov @chhagedorn @TobiHartmann

**Pull request closed without merging**